app/src/main/java/lu/geoportail/map/LuxTileServer.kt

```diff
@@ -35,71 +47,241 @@ class LuxTileServer(
+            if(!toFile.parentFile.exists()) Files.createDirectories(toFile.parentFile.toPath())
+            File(fromPath, filename).renameTo(toFile)
+        }
+    }
+
+    private fun getPaths(ressourceName: String): JSONArray {
+        // metadata will be read from local file until it is available online
+        return try {
+            val url = URL("file:${this.filePath}/dl/versions/$ressourceName.meta")
+            val json = JSONObject(url.openStream().use { it.readBytes() }.toString(Charsets.UTF_8))
+            val paths = json.getJSONArray("sources") ?: JSONArray()
+            paths.put("file:/versions/$ressourceName.meta")
+        } catch (e: FileNotFoundException) {
+            JSONArray()
+        }
+    }
+
+    private fun getVer(ressourceName: String): String? {
+        return try {
+            val url = URL("file:${this.filePath}/dl/versions/$ressourceName.meta")
+            val json = JSONObject(url.openStream().use { it.readBytes()}.toString(Charsets.UTF_8))
+            json.getString("version")
+        }
+        catch (e: FileNotFoundException) {
+            null
+        }
+    }
+
+    private fun getSize(ressourceName: String): String? {
+        return try {
+            val dlPath = when(getStatus(ressourceName)) {
+                "in progress" -> "tmp"
+                else -> "dl"
+            }
+            val url = URL("file:${this.filePath}/$dlPath/versions/$ressourceName.meta")
+            val json = JSONObject(url.openStream().use { it.readBytes()}.toString(Charsets.UTF_8))
+            val src = json.getJSONArray("sources")
+            var totalSize: Long = 0
+            for (i in 0 until src.length()) {
+                try {
+                    val filename = URL(src.getString(i)).file
+                    totalSize += Files.size(
+                        File(
+                            File(this.filePath, dlPath),
+                            filename
+                        ).toPath()
+                    )
+                }
+                catch (e: NoSuchFileException) {}
+            }
+            totalSize.toString()
+        }
+        catch (e: FileNotFoundException) {
+            null
+        }
+    }
+
+    private fun getStatus(resName: String): String {
+        val dl = dlThreads[resName]
+        dl?.join(1)
+        return when {
+            dl?.isAlive() ?: false -> "in progress"
+            dl?.state == Thread.State.TERMINATED -> "done"
+            else -> "undefined"
+        }
+    }
+
+    private fun saveMeta(meta: JSONObject, pathName: String) {
+        // metadata will be read from local file until it is available online
+        val parentFolder = File(pathName).parentFile
+        if(!parentFolder.exists()) Files.createDirectories(parentFolder.toPath())
+        val outputStream = FileOutputStream(File(pathName))
+        outputStream.use { it.write(meta.toString().toByteArray())}
+    }
+
+    private fun getAllMeta(): JSONObject? {
```

Comment:
Can you memoize this function instead of downloading it each time?

Reply:
I have to think it over, but since this is the core of the computation of update status, it must not be cached for too long. Currently the file is downloaded only once for check and once for update fonctions. This makes sure it is up to date.